Python/optimizer_bytecodes.c

```diff
@@ -686,6 +686,32 @@ dummy_func(void) {
         }
     }
 
+    op(_GUARD_CALL_BUILTIN_O, (callable, self_or_null, unused[oparg] -- callable, self_or_null, unused[oparg])) {
+        int total_args = oparg;
+        if (sym_is_null(self_or_null) || sym_is_not_null(self_or_null)) {
+            total_args += sym_is_not_null(self_or_null);
+            // Constant propagate
+            if (total_args == 1 && sym_is_const(ctx, callable)) {
+                PyObject *callable_o = sym_get_const(ctx, callable);
+                if (PyCFunction_CheckExact(callable_o) &&
+                    PyCFunction_GET_FLAGS(callable_o) == METH_O) {
+                    REPLACE_OP(this_instr, _NOP, 0 ,0);
+                }
+            }
+        }
```

Comment:
Not sure how much this helps in practice, but in theory we could infer a `NULL`/non-`NULL` here based on the oparg, and narrow the type of the callable:
```suggestion
        if (sym_is_null(self_or_null) || sym_is_not_null(self_or_null)) {
            int total_args = oparg + sym_is_not_null(self_or_null);
            // Constant propagate
            if (total_args == 1 && sym_is_const(ctx, callable)) {
                PyObject *callable_o = sym_get_const(ctx, callable);
                if (PyCFunction_CheckExact(callable_o) &&
                    PyCFunction_GET_FLAGS(callable_o) == METH_O) {
                    REPLACE_OP(this_instr, _NOP, 0 ,0);
                }
            }
        }
        else if (oparg == 0) {
            sym_set_non_null(ctx, self_or_null);
        }
        else {
            assert(oparg == 1);
            sym_set_null(ctx, self_or_null);
        }
        sym_set_type(callable, &PyCFunction_Type);
```